### PR TITLE
Changed inner_hits to work with the new join field type and

### PR DESCRIPTION
**Pull request by martijnvg**
at the same time maintaining support for the `_parent` meta field type.

Relates to #20257

### CODE REVIEW

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/ParentChildInnerHitContextBuilder.java [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+            TopDocs[] result = new TopDocs[hits.length];
+            for (int i = 0; i < hits.length; i++) {
+                SearchHit hit = hits[i];
+                SearchHitField joinField = hit.getFields().get(joinFieldMapper.name());
+                if (joinField == null) {
+                    result[i] = Lucene.EMPTY_TOP_DOCS;
+                    continue;
+                }
+
+                String parentIdFieldName = ParentJoinFieldMapper.getParentIdFieldName(joinFieldMapper.name(), typeName);
```
Comment (jimczi) [inline modules/parent-join/src/main/java/org/elasticsearch/join/query/ParentChildInnerHitContextBuilder.java]: I think we can remove the leniency and make sure that we build `parent` or `children` inner hits depending on the context. `hasParent` should only build inner hits of `parent` documents and hasChild for `children` documents.
With this information in `JoinFieldInnerHitSubContext` you can just call:
````
joinFieldMapper.getJoinFieldMapper(typeName, isParentInnerHits);
````
and use `JoinFieldMapper#hasParent` and `JoinFieldMapper#hasChild` to build the inner hits query ?

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/ParentChildInnerHitContextBuilder.java [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+                SearchHitField parentIdField = hit.getFields().get(parentIdFieldName);
+                QueryShardContext qsc = context.getQueryShardContext();
+
+                Query q;
+                if (parentIdField == null) {
```
Comment (jimczi) [inline modules/parent-join/src/main/java/org/elasticsearch/join/query/ParentChildInnerHitContextBuilder.java]: Does this work if the document is a child **and** a parent ? I think it's simpler if we use the context directly.
`hasParent` expects a `child` and builds `parent` inner hits. `hasChildren` expects a `parent` and builds `children` inner hits ?

**Review by jimczi — COMMENTED** [review]
Thank you for rewriting all the tests. 
I left one comment. I think we need to be explicit when we build the inner hit query.

**Comment by martijnvg** [comment]
@jimczi I made the join field inner hits logic more explicit.

**Review by jimczi — APPROVED** [review]
LGTM

**Comment by Yneth** [comment]
My mapping
`{
  "mappings": {
    "test": {
      "properties": {
        "views": {
          "type": "nested",
          "properties": {
            "date": {
              "type": "date"
            }
          }
        }
      }
    }
  }
}`

My query to sort by view count
`{
    "query": {
        "has_child": {
            "type": "views",
            "score_mode": "sum",
            "query": {"match_all": {}}, // here I would also like to add some filtering by date
            "inner_hits": {}
        }
    }
}`

throws 
`{
    "error": {
        "root_cause": [
            {
                "type": "query_shard_exception",
                "reason": "[has_child] no join field has been configured",
                "index_uuid": "XW9AAGtjRtutQ3n4mhZ9zQ",
                "index": "test"
            }
        ],
        "type": "search_phase_execution_exception",
        "reason": "all shards failed",
        "phase": "query",
        "grouped": true,
        "failed_shards": [
            {
                "shard": 0,
                "index": "test",
                "node": "_bbwp3g_TfOGbZE4McKnjw",
                "reason": {
                    "type": "query_shard_exception",
                    "reason": "[has_child] no join field has been configured",
                    "index_uuid": "XW9AAGtjRtutQ3n4mhZ9zQ",
                    "index": "test"
                }
            }
        ]
    },
    "status": 400
}
`

**Pull request merged**